src/plugins/es_ui_shared/static/forms/hook_form_lib/hooks/use_field.ts

```diff
@@ -0,0 +1,360 @@
+  };
+
+  const onValueChange = () => {
+    setIsChangingValue(true);
+
+    // Update the form data observable
+    form.__updateFormDataAt(path, serializeOutput(value));
+
+    // Validate field(s) and set form.isValid flag
+    form.__validateFields(fieldsToValidateOnChange);
+
+    if (debounceTimeout.current) {
+      clearTimeout(debounceTimeout.current);
+    }
+
+    debounceTimeout.current = setTimeout(() => {
+      setIsChangingValue(false);
+    }, errorDisplayDelay);
+  };
+
+  const runValidations = ({
+    formData,
+    value: valueToValidate,
+    validationTypeToValidate,
+  }: {
+    formData: any;
+    value: unknown;
+    validationTypeToValidate?: string;
+  }): ValidationError[] | Promise<ValidationError[]> => {
+    let validationErrors: ValidationError[] = [];
+    let hasToRunAsync = false;
+
+    // By default, for fields that have an asynchronous validation
+    // we will clear the errors as soon as the field value changes.
+    clearErrors([VALIDATION_TYPES.FIELD, VALIDATION_TYPES.ASYNC]);
+
+    const runAsync = async () => {
+      for (const validation of validations) {
+        const {
+          validator,
+          exitOnFail = true,
+          type: validationType = VALIDATION_TYPES.FIELD,
+        } = validation;
+
+        if (
+          typeof validationTypeToValidate !== 'undefined' &&
+          validationType !== validationTypeToValidate
+        ) {
+          continue;
+        }
+
+        const validationResult = await validator({
+          value: (valueToValidate as unknown) as string,
+          errors: validationErrors,
+          form,
+          formData,
+          path,
+        });
+
+        if (!validationResult) {
+          continue;
+        }
+
+        validationErrors.push({
+          ...validationResult,
+          validationType: validationType || VALIDATION_TYPES.FIELD,
+        });
+
+        if (exitOnFail) {
+          break;
+        }
+      }
+
+      return validationErrors;
+    };
+
+    const runSync = () => {
+      // Sequencially execute all the validations for the field
+      for (const validation of validations) {
+        const {
+          validator,
+          exitOnFail = true,
+          type: validationType = VALIDATION_TYPES.FIELD,
+        } = validation;
+
+        if (
+          typeof validationTypeToValidate !== 'undefined' &&
+          validationType !== validationTypeToValidate
+        ) {
+          continue;
+        }
+
+        const validationResult: ValidationError = validator({
+          value: (valueToValidate as unknown) as string,
+          errors: validationErrors,
+          form,
+          formData,
+          path,
+        }) as ValidationError;
+
+        if (!validationResult) {
+          continue;
+        }
+
+        if (!!validationResult.then) {
+          // The validator returned a Promise.
+          // Abort and run the validations asynchronously
+          hasToRunAsync = true;
+          break;
```

Comment:
I think this can be simplified a bit by removing `hasToRunAsync` (haven't tested this though).

```js
        if (!!validationResult.then) {
          // The validator returned a Promise.
          // Abort and run the validations asynchronously
          return runAsync();
        }
```

We also don't need to worry about clearing `validationErrors` if each helper defines it in its own scope.

```ts
    const runAsync = async () => {
      const validationErrors: ValidationError[] = [];
      /* ... */
    }

    const runSync = () => {
      const validationErrors: ValidationError[] = [];
      /* ... */
    }
```